### ISSUE TITLE
Possible inefficiency with step size in Newton method + backtracking linesearch

### ISSUE THREAD

**Issue by sebastien-villemot** [JuliaOpt/Optim.jl#91]
I noticed that the Newton method is implemented in such a way that the step size (as embodied in the parameter `alpha`) is not reset across iterations. This affects at least the backtracking linesearch.

In practice, if at some iteration the step size as decided by the line search is very small, then at the next iteration, the line search will start with the previous value for `alpha`, instead of starting with one. So the parameter `alpha` can only decrease across iterations when using the backtracking linesearch.

This can be highly inefficient for some problems, resulting in many more iterations than necessary.

Is this a deliberate choice? In my `NLsolve` package, I chose to always start linesearches with `alpha=1`, since it results in much more efficients outcomes for the problems that I tested: https://github.com/EconForge/NLsolve.jl/commit/b14786644a03e9e45de7dd61277839110fab482b

**Comment by matthieugomez** [comment]
In my problem, `NLsolve` returns the error "Too many iterations in backtracking_linesearch!". So it does not seem like a good choice in general.

**Comment by mlubin** [comment]
PRs accepted

**Comment by johnmyleswhite** [comment]
Arguably we should remove backtracking_linesearch since it never works well in my experience.

**Comment by johnmyleswhite** [comment]
Any progress on this?

**Comment by pkofod** [comment]
Should we perhaps deprecate backtracking_linesearch?

**Comment by pkofod** [comment]
I think we should benchmark the change before doing anything, but resetting to 1.0 seems like the right thing to do. I have other stuff to finish, and limited time, but if no one else grabs it, I'll have a look at some point.

**Comment by cortner** [comment]
please don't remove backtracking line search; I am just in the process of rewriting it. On my first tests, I actually beat HZ and MT line search with LBFGS.

**Comment by cortner** [comment]
Btw, I just notices that the same 1.0 problem is also in LBFGS.

**Comment by cortner** [comment]
Personally, I always use an extrapolation step, which defaults to 1.0 if it ends up near 1.0. This is my current implementation; just as a quick hack to try it out:
```julia
        if iteration > 1
           alphaguess = 2.0 * (f_x - f_x_previous) / dphi0
           alphaguess = max(alphaguess, alpha/4.0)
           alphaguess = min(alphaguess, 1.0)
           if alphaguess > 0.75
               alphaguess = 1.0
           end
        else
            alphaguess = 1.0
        end
```

On Rosenbrock example: 

linesearch | F-evals | G-evals
HZ         |    60    |   81 
MT        |    76    |  76
new BT  |    45   |  72 

(HZ is for ConjugateGradients, the rest for LBFGS)

**Comment by cortner** [comment]
some variant of this might even fix #166

**Comment by pkofod** [comment]
@cortner where are we on this, was this fixed somewhere?

**Comment by anriseth** [comment]
An improvement to backtracking was introduced as a non-default option here: anriseth/LineSearches.jl#7.
Further updates that change default behaviour of backtracking will be merged with anriseth/LineSearches.jl#20.

The Newton algorithm still seems to be lacking the resetting `state.alpha = 1.0` step, which is the best default in the "good" regions for Newton.

**Comment by pkofod** [comment]
But this is now a LineSearches.jl issue right? Or do you need something from optim ?

**Comment by cortner** [comment]
No - it is more complicated. What Linesearches.jl really needs is a refactoring to make linesearches tyoes rather than functions: this will definitely affect Optim.

**Comment by cortner** [comment]
See #294  and LineSearches.jl:#9

**Comment by pkofod** [comment]
closed by #328
